includes/validation/class-amp-site-validation.php

```diff
@@ -0,0 +1,154 @@
+	 * @return string[] $permalinks The post permalinks in an array.
+	 */
+	public static function get_post_permalinks( $post_type, $number_posts = 200, $offset = null ) {
+		$args = array(
+			'post_type'      => $post_type,
+			'posts_per_page' => $number_posts,
+			'post_status'    => 'publish',
+			'orderby'        => 'ID',
+			'order'          => 'ASC',
+			'fields'         => 'ids',
+		);
+		if ( is_int( $offset ) ) {
+			$args = array_merge( $args, compact( 'offset' ) );
+		}
+		$query = new WP_Query( $args );
+
+		return array_map( 'get_permalink', $query->posts );
+	}
+
+	/**
+	 * Gets the front-end links for public taxonomy terms, like categories and tags.
+	 *
+	 * For example, https://example.org/?cat=2
+	 * This includes categories and tags, and any more that are registered.
+	 * But it excludes post_format terms.
+	 *
+	 * @param string $taxonomy     The name of the taxonomy, like 'category' or 'post_tag'.
+	 * @param int    $number_links The maximum amount of links to get (optional).
+	 * @param int    $offset       The number at which to offset the query (optional).
+	 * @return string[] $links The term links in an array.
+	 */
+	public static function get_taxonomy_links( $taxonomy, $number_links = 200, $offset = null ) {
+		$args = array(
+			'taxonomy' => $taxonomy,
+			'orderby'  => 'id',
+			'number'   => $number_links,
+		);
+		if ( is_int( $offset ) ) {
+			$args = array_merge( $args, compact( 'offset' ) );
+		}
+
+		return array_map( 'get_term_link', get_terms( $args ) );
+	}
+
+	/**
+	 * Validates the URLs of the entire site.
+	 *
+	 * Accepts an optional parameter of a WP-CLI progress bar object.
+	 * Calling its tick() method updates the display in WP-CLI to show the percentage of the site crawl that's complete.
+	 *
+	 * @todo: Consider wrapping this function with another, as different use cases will probably require a different return value or display.
+	 * For example, the <button> in /wp-admin that makes an AJAX request for this will need a different response than a WP-CLI command.
+	 * @param object $wp_cli_progress The object that shows progress in the WP-CLI script to validate the site.
+	 * @return array $urls_validated The URLs that were visited, regardless of what the validation result was.
+	 */
+	public static function validate_entire_site_urls( $wp_cli_progress = null ) {
+		// Validate the homepage.
+		self::validate_urls( home_url( '/' ) );
+
+		// Validate all public, published posts.
+		$public_post_types = get_post_types( array( 'public' => true ), 'names' );
+		foreach ( $public_post_types as $post_type ) {
+			$permalinks = self::get_post_permalinks( $post_type, self::BATCH_SIZE );
+			$offset     = 0;
+
+			while ( ! empty( $permalinks ) ) {
+				self::validate_urls( $permalinks );
+				$offset    += self::BATCH_SIZE;
+				$permalinks = self::get_post_permalinks( $post_type, self::BATCH_SIZE, $offset );
+			}
+
+			if ( $wp_cli_progress ) {
+				$wp_cli_progress->tick();
```

Comment:
This could get de-coupled from WP-CLI by passing a `$progress_callback` function instead which is called, and then you could call the method like so:

```php
\AMP_Site_Validation::validate_entire_site_urls( function() use ( $wp_cli_progress ) {
    $wp_cli_progress->tick()
} ) 
```